packages/near-fast-auth-signer/src/components/SignMultichain/utils.ts

```diff
@@ -0,0 +1,239 @@
+import { formatEther } from 'ethers';
+import hash from 'hash.js';
+import keccak from 'keccak';
+import { base_decode } from 'near-api-js/lib/utils/serialize';
+
+import { ethereumSchema } from './ethereum/schema';
+// eslint-disable-next-line import/no-cycle
+import { getEthereumGasFee, getEthereumMessageToSign } from './ethereum/sign';
+import { MultichainIframeMessage, DerivationPathDeserialized } from './types';
+import { fetchGeckoPrices } from '../Sign/Values/fiatValueManager';
+
+// TODO: use this for blacklisting on limited access key creation AND sign
+const MULTICHAIN_CONTRACT_TESTNET = 'multichain-testnet-2.testnet';
+const MULTICHAIN_CONTRACT_MAINNET = 'multichain-testnet-2.testnet';
+
+export const getMultiChainContract = () => (process.env.NETWORK_ID === 'mainnet' ? MULTICHAIN_CONTRACT_MAINNET : MULTICHAIN_CONTRACT_TESTNET);
+
+const getSchema = (asset: DerivationPathDeserialized['asset']) => {
+  switch (asset) {
+    case 'ETH':
+      return ethereumSchema;
+    default:
+      return null;
+  }
+};
+
+export const validateMessage = async (message: MultichainIframeMessage, asset: DerivationPathDeserialized['asset']): Promise<boolean
+| Error> => {
+  const schema = getSchema(asset);
+  if (!schema) {
+    throw new Error(`Schema for asset ${asset} is not defined`);
+  }
+
+  try {
+    await schema.validate(message);
+    return true;
+  } catch (e) {
+    throw new Error(e);
+  }
+};
+
+export const najPublicKeyStrToUncompressedHexPoint = (najPublicKeyStr) => `04${base_decode(najPublicKeyStr.split(':')[1]).toString('hex')}`;
+
+export const sha256Hash = async (str) => {
+  const encoder = new TextEncoder();
+  const data = encoder.encode(str);
+
+  const hashBuffer = await crypto.subtle.digest('SHA-256', data);
+
+  const hashArray = Array.from(new Uint8Array(hashBuffer));
+  return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
+};
+
+export const sha256StringToScalarLittleEndian = (hashString) => {
+  const littleEndianString = hashString.match(/../g).reverse().join('');
+
+  const scalar = new BN(littleEndianString, 16);
+
+  return scalar;
+};
+
+export const deriveChildPublicKey = async (
+  parentUncompressedPublicKeyHex,
+  signerId,
+  path = ''
+) => {
+  const ec = new EC('secp256k1');
+  const scalar = await sha256Hash(
+    `near-mpc-recovery v0.1.0 epsilon derivation:${signerId},${path}`
+  );
+  const scalarLittleEndian = sha256StringToScalarLittleEndian(scalar);
+
+  const x = parentUncompressedPublicKeyHex.substring(2, 66);
+  const y = parentUncompressedPublicKeyHex.substring(66);
+
+  // Create a point object from X and Y coordinates
+  const oldPublicKeyPoint = ec.curve.point(x, y);
+
+  // Multiply the scalar by the generator point G
+  const scalarTimesG = ec.g.mul(scalarLittleEndian);
+
+  // Add the result to the old public key point
+  const newPublicKeyPoint = oldPublicKeyPoint.add(scalarTimesG);
+
+  return `04${
+    newPublicKeyPoint.getX().toString('hex')
```

Comment:
please update this with https://pagodaplatform.slack.com/archives/C04HWL4BJKE/p1709672855903079?thread_ts=1709638154.085769&cid=C04HWL4BJKE